parlai/core/torch_generator_agent.py

```diff
@@ -1750,27 +1730,33 @@ def select_paths(self, logprobs, prior_scores, current_length) -> _PathSelection
         # get the actual word id from residual of the same division
         tok_ids = best_idxs % voc_size
 
-        tok_scores, tok_ranks = None, None
+        token_details: Optional[List[_PathSelectionTokenDetails]] = None
         if self.verbose:
+            probs = torch.softmax(logprobs, dim=-1)
             tok_scores = (
-                torch.index_select(logprobs, 0, hyp_ids)
+                torch.index_select(probs, 0, hyp_ids)
                 .gather(1, tok_ids.unsqueeze(1))
                 .view(-1)
             )
-
             tok_ranks = (
-                logprobs.argsort(1, descending=True)
+                probs.argsort(1, descending=True)
                 .argsort(1)
                 .view(-1)
                 .gather(0, best_idxs)
             )
 
+            token_details = []
+
+            for score, rank in zip(tok_scores.cpu().numpy(), tok_ranks.cpu().numpy()):
+                token_details.append(
+                    {"token_score": score.item(), "token_rank": int(rank.item())}
```

Comment:
To my understanding, TypedDict's are really just dicts that the Python typing system has some extra information about (i.e. you implement them just as dicts), and I typed `token_details` on line 1733. 

From the PEP:
> Methods are not allowed, since the runtime type of a TypedDict object will always be just dict (it is never a subclass of dict).

Reply:
This looks to be the correct approach from the PEP, and the type system is happy. From the PEP:

> Another example where a type checker should treat a dictionary display as a TypedDict is in an assignment to a variable with a previously declared TypedDict type
```
class Movie(TypedDict):
    name: str
    year: int

movie: Movie
...
movie = {'name': 'Blade Runner', 'year': 1982}